service/udp.go

```diff
@@ -321,177 +449,231 @@ func (c *natconn) ReadFrom(buf []byte) (int, net.Addr, error) {
-	}
-
+// Del deletes a UDP NAT entry from the natmap.
+func (m *natmap) Del(clientAddr string) {
 	m.Lock()
 	defer m.Unlock()
 
-	m.keyConn[key] = entry
-	return entry
+	if _, ok := m.associations[clientAddr]; ok {
+		delete(m.associations, clientAddr)
+	}
 }
 
-func (m *natmap) del(key string) net.PacketConn {
+// Add adds a new UDP NAT entry to the natmap.
+func (m *natmap) Add(clientAddr string, assoc PacketAssociation) {
 	m.Lock()
 	defer m.Unlock()
 
-	entry, ok := m.keyConn[key]
-	if ok {
-		delete(m.keyConn, key)
-		return entry
-	}
-	return nil
-}
-
-func (m *natmap) Add(clientAddr net.Addr, clientConn net.PacketConn, cryptoKey *shadowsocks.EncryptionKey, targetConn net.PacketConn, keyID string) *natconn {
-	connMetrics := m.metrics.AddUDPNatEntry(clientAddr, keyID)
-	entry := m.set(clientAddr.String(), targetConn, cryptoKey, connMetrics)
-
-	go func() {
-		timedCopy(clientAddr, clientConn, entry, m.logger)
-		connMetrics.RemoveNatEntry()
-		if pc := m.del(clientAddr.String()); pc != nil {
-			pc.Close()
-		}
-	}()
-	return entry
+	m.associations[clientAddr] = assoc
 }
 
 func (m *natmap) Close() error {
 	m.Lock()
 	defer m.Unlock()
 
 	var err error
-	now := time.Now()
-	for _, pc := range m.keyConn {
-		if e := pc.SetReadDeadline(now); e != nil {
+	for _, assoc := range m.associations {
+		if e := assoc.Close(); e != nil {
 			err = e
 		}
 	}
 	return err
 }
 
-// Get the maximum length of the shadowsocks address header by parsing
-// and serializing an IPv6 address from the example range.
-var maxAddrLen int = len(socks.ParseAddr("[2001:db8::1]:12345"))
+// PacketHandleFunc processes a single incoming packet.
+type PacketHandleFunc func(pkt []byte, assoc PacketAssociation) error
+
+// PacketHandleFuncWithLazySlice processes a single incoming packet.
+//
+// lazySlice is the LazySlice that holds the pkt buffer, which should be
+// released as soon as the packet is processed.
+type PacketHandleFuncWithLazySlice func(pkt []byte, assoc PacketAssociation, lazySlice slicepool.LazySlice)
+
+func HandleAssociation(assoc PacketAssociation, handle PacketHandleFuncWithLazySlice) {
+	for {
+		lazySlice := readBufPool.LazySlice()
+		buf := lazySlice.Acquire()
+		n, err := assoc.ReadFromClient(buf)
+		if errors.Is(err, net.ErrClosed) {
+			lazySlice.Release()
+			return
+		}
+		pkt := buf[:n]
+		select {
+		case <-assoc.Done():
+			lazySlice.Release()
+			return
+		default:
+			go handle(pkt, assoc, lazySlice)
+		}
+	}
+}
 
-// copy from target to client until read timeout
-func timedCopy(clientAddr net.Addr, clientConn net.PacketConn, targetConn *natconn, l *slog.Logger) {
-	// pkt is used for in-place encryption of downstream UDP packets, with the layout
-	// [padding?][salt][address][body][tag][extra]
+// HandleAssociationTimedCopy handles the target-side of the association by
+// copying from target to client until read timeout.
+func HandleAssociationTimedCopy(assoc PacketAssociation, handle PacketHandleFunc) {
```

Comment:
I think something like RelayTargetToClient would be more explicit.

But the signature is confusing. There's a modeling issue. Where's the target in the function signature? That needs to be more explicit.

The association is an association with the client, not the target. It represents a (client address, server address) tuple. The handler takes that association, authenticates and creates a single target PacketConn to talk to the targets.

The targets are related to handling the association.